log/loki.go

```diff
@@ -0,0 +1,499 @@
+			oldLogs = oldLogs[:len(oldLogs)+oldCount]
+
+			t := time.Now()
+			cutOffIndex := sortAndSplitMsgs(oldLogs, bufferTime)
+			if cutOffIndex == 0 {
+				continue
+			}
+			t1 := time.Since(t)
+
+			strms := h.createPushMessage(oldLogs, cutOffIndex, oldDropped)
+			if cutOffIndex > len(oldLogs) {
+				oldLogs = oldLogs[:0]
+				continue
+			}
+			oldLogs = oldLogs[:copy(oldLogs, oldLogs[cutOffIndex:])]
+			t2 := time.Since(t) - t1
+
+			var b bytes.Buffer
+			_, err := strms.WriteTo(&b)
+			if err != nil {
+				fmt.Printf("Error while marshaling logs for loki %s\n", err)
+				continue
+			}
+			size := b.Len()
+			t3 := time.Since(t) - t2 - t1
+
+			err = h.push(b)
+			if err != nil {
+				fmt.Printf("Error while sending logs to loki %s\n", err)
+				continue
+			}
+			t4 := time.Since(t) - t3 - t2 - t1
+
+			if h.profile {
+				fmt.Printf("sorting=%s, adding=%s marshalling=%s sending=%s count=%d final_size=%d\n",
+					t1, t2, t3, t4, cutOffIndex, size)
+			}
+		}
+	}()
+
+	for {
+		select {
+		case entry := <-h.ch:
+			if count == h.limit {
+				dropped++
+				continue
+			}
+
+			// Arguably we can directly generate the final marshalled version of the labels right here
+			// through sorting the entry.Data, removing additionalparams from it and then dumping it
+			// as the final marshal and appending level and h.labels after it.
+			// If we reuse some kind of big enough `[]byte` buffer we can also possibly skip on some
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
+			msgs[count] = tmpMsg{
+				labels: labels,
+				msg:    entry.Message,
+				t:      entry.Time.UnixNano(),
+			}
+			count++
+		case <-ticker.C:
+			ch := make(chan time.Duration)
+			pushCh <- ch
+			ch <- h.pushPeriod / 2
+			<-ch
+		case <-h.ctx.Done():
+			ch := make(chan time.Duration)
+			pushCh <- ch
+			ch <- 0
+			<-ch
+			return
+		}
+	}
+}
+
+func sortAndSplitMsgs(msgs []tmpMsg, bufferTime time.Duration) int {
+	if len(msgs) == 0 {
+		return 0
+	}
+
+	// TODO using time.Before was giving a lot of out of order, but even now, there are some, if the
+	// limit is big enough ...
+	sort.Slice(msgs, func(i, j int) bool {
+		return msgs[i].t < msgs[j].t
+	})
+
+	// We can technically cutoff during the addMsg phase, which will be even better if we sort after
+	// it as well ... maybe
+	cutOff := time.Now().Add(-bufferTime).UnixNano() // probably better to be configurable
```

Comment:
Instead of passing `bufferTime` as a `time.Duration`, can't you pass it as `time.Time` directly? You can get it from the [`ticker.C` channel](https://tip.golang.org/pkg/time/#Ticker), and you wouldn't have to call `time.Now()` here, which can after all slightly drift on a CPU-heavy test with a lot of logs, while you're sorting them above.